Purchases/Purchasing/ProductInfo.swift

```diff
@@ -69,54 +39,60 @@ struct ProductInfo {
             return key
         }
 
-        if #available(iOS 12.2, macOS 10.14.4, tvOS 12.2, *) {
-            for offer in discounts {
-                key += "-\(offer.offerIdentifier ?? "null offer id")"
-            }
+        for offer in discounts {
+            key += "-\(offer.offerIdentifier ?? "null offer id")"
         }
         return key
     }
 
-    private func discountsAsDictionary() -> [String: NSObject] {
-        var discountDict: [String: NSObject] = [:]
-            if #available(iOS 12.2, macOS 10.14.4, tvOS 12.2, *) {
-                if let discounts = self.discounts {
-                    let offers = NSMutableArray()
-                    for discount in discounts {
-                        guard let offerIdentifier = discount.offerIdentifier else {
-                            break
-                        }
-
-                        offers.add(["offer_identifier": offerIdentifier,
-                                    "price": discount.price,
-                                    "payment_mode": discount.paymentMode.rawValue])
-
-                    }
-                    discountDict["offers"] = offers
-                }
-            }
-        return discountDict
+}
+
+extension ProductInfo: Encodable {
+
+    public enum CodingKeys: String, CodingKey {
+
+        case productIdentifier = "product_id"
+        case paymentMode = "payment_mode"
+        case currencyCode = "currency"
+        case price
+        case normalDuration = "normal_duration"
+        case introDuration = "intro_duration"
+        case trialDuration = "trial_duration"
+        case introPrice = "introductory_price"
+        case subscriptionGroup = "subscription_group_id"
+        case discounts = "offers"
+
     }
 
-    private func productDurationsAsDictionary() -> [String: NSObject] {
-        var durations: [String: NSObject] = [:]
+    func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
 
-        if let normalDuration = normalDuration as NSString? {
-            durations["normal_duration"] = normalDuration
-        }
+        try container.encode(self.productIdentifier, forKey: .productIdentifier)
 
-        guard let introDuration = introDuration as NSString? else {
-            return durations
+        if self.paymentMode != .none {
+            try container.encode(self.paymentMode, forKey: .paymentMode)
         }
+        try container.encode(self.currencyCode, forKey: .currencyCode)
+        try container.encode((self.price as NSDecimalNumber).description, forKey: .price)
```

Comment:
let's add a comment explaining why we're using strings here to prevent precision issues

Reply:
Yup, done!